### PR TITLE
fix(chat): improve tool group summaries and scrolling

### PR DESCRIPTION
**Pull request by juliusmarminge**
## Summary

Tool-group summaries showed raw T3 tool names while running, counted browser actions as generic tools, and let long expanded lists fill the conversation.

- Share lifecycle-aware labels across the summary and rows, including "Clicking", "Clicked", and explicit failure states. Keep the summary on any still-running call and stop its shimmer once none remain active.
- Count browser actions separately, including mixed summaries such as "Ran 4 commands and used browser 15 times".
- Use globe icons for browser activity and add icons to every row. Non-browser T3 tools keep their existing icon.
- Extract the actual command from bash and zsh wrappers while preserving the full command in details.
- Bound and virtualize expanded groups, with internal scrolling and edge fades on web/desktop and mobile. Short groups still fit their contents.
- Preserve expanded details and row-relative scroll anchors across collapse/reopen and virtualization. Follow only genuine appends at the hard end, not status updates or disclosure changes.
- Use the mobile 14px rem for fixed-height estimates, trust measured content height, and remeasure native text when typography changes. Wait for viewport layout when a short group grows.

This changes client presentation only. No server or wire-contract changes.

## Verification

- 371 focused tests passed across shared command/presentation logic, web timeline/session logic, and mobile thread activity, layout, and live-follow behavior.
- Web and shared-runtime typechecks, mobile `tsc`, targeted lint, and `git diff --check` passed.
- Browser verification covered running and completed groups, scroll fades, row icons, bash/zsh details, and mixed browser counts.
- iOS Simulator verification covered nested gestures, fast flicks, expanded output, append following, collapse/reopen, and Dynamic Type in both directions. Android and physical-device frame timing remain unverified.

## Full recording

Complete 31-second walkthrough, cropped to the conversation. This uses seeded UI states, not a live provider run.

https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/d150322651a22284/tool-groups-full-recording.mp4

[Open the full recording](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/d150322651a22284/tool-groups-full-recording.mp4)

## Scroll regression verification

[Full scroll verification recording](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/7b225fbf86d8760e/tool-group-scroll-verification.mp4) — 52 seconds, conversation-only crop of seeded web states.

- Expanded-result restoration retained the same 288px scroll offset and -120px row-relative offset after reopening.
- Appending while reading older calls caused 0px movement; appending at the hard end kept the end gap at 0px. Status changes and an interrupted append were checked separately.
- Forced outer-list recycling preserved the expanded result and its row-relative position. A group with 1,000 additional calls stayed capped at 288px with 23–33 mounted rows in the sampled positions.
- A one-time web load check reconciles a reproduced case where the list reported a restored offset but the DOM remained at 0px. Fault injection confirmed the repair from 0px to 288px; it does not run during normal scrolling or data updates.

These web checks cover position, virtualization, and follow behavior, not a frame-time benchmark. Native results are below.

## Native scroll verification

iPhone 17 Pro, iOS 26.5 Simulator, using the installed development client against isolated seeded state. No provider commands were run for the injected updates.

The pass caught and fixed a fast-flick restoration jump, incorrect mobile row-height estimates, and stale native text measurements. Web and mobile now share the offset-based anchor lookup instead of trusting the virtualizer's cached visible index.

- Slow scrolling, fast flicks, and scrolling inside expanded output left the outer conversation position unchanged in the checked gestures.
- Appending and completing a call while reading older rows moved neither list. Appending at the hard end advanced one row and retained a zero end gap.
- Collapse/reopen retained the expanded result and its 112.67pt row-relative offset. Flicking to the top now restores row zero instead of jumping five rows down.
- Dynamic Type changed row height from 29pt to 51.25pt while retaining the 256pt viewport and the same row. Its 22pt intra-row offset returned exactly after restoring normal text, with less than 0.1pt rounding at the larger size. Collapsed headers remained readable too.

Typography changes alone clear the outer size cache and remount group content/header text. The outer row and pressable hosts remain mounted. Scrolling, streamed output, and disclosure toggles do not trigger that invalidation.

Full native gesture, append, and restoration walkthrough, 107 seconds:

https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/a9cc0cd3dca0c60f/native-scroll-verification.mp4

Final text-size round trip, 19 seconds:

https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/c4a84bdc8933c459/native-dynamic-type-verification.mp4

| Native fix | Before and after |
| --- | --- |
| Restore the actual top row after a fast flick | Before: reopening jumps to the sixth call. After: the first call stays at the top. |
| Remeasure accessibility text instead of clipping it | Before: enlarged text is vertically clipped. After: tool rows and summaries fit their text. |

This is behavioral verification in Simulator, not an FPS benchmark or an Android/physical-device test. The recordings are cropped above the composer.

## Before and after

Before images are the reported regressions. After images are focused frames from the verified web walkthrough.

| Fix | Before | After |
| --- | --- | --- |
| Friendly running labels and correct tense | Before: raw browser tool name in the running summary | Running<br>After: Clicking in the preview browser<br>Completed<br>After: Clicked in the preview browser |
| Globe icons on every browser row | Before: only the failed browser call has a leading T3 icon | After: browser rows use globes, other T3 tools retain their icon, and failures remain marked |
| Bounded scrolling with edge fades | Before: the expanded tool list fills the conversation | After: the tool list scrolls within a bounded region with fades above and below |
| Actual process names inside shell wrappers | Before: the summary says Running zsh | After: Running vp with both bash and zsh wrappers preserved in details |
| Browser-specific and mixed counts | Before: browser-only activity says Used 18 tools | After: separate browser counts in browser-only and mixed groups |

Created in T3 Code.

---

> [!NOTE]
> **Medium Risk**
> Large UI/scroll behavior changes across mobile and web timelines (virtualization, nested scroll, live-follow), with broad test coverage but high regression surface for scrolling and tool-group presentation.
> 
> **Overview**
> Improves how expanded tool-call groups behave in the thread timeline on **web and mobile**: long groups scroll inside a capped, virtualized list with edge fades instead of stretching the whole conversation, and **scroll position plus open row details** survive collapse, list recycling, and thread switches via row-relative anchors shared with `client-runtime`.
> 
> **Presentation** moves to shared `resolveWorkEntryToolPresentation` logic: lifecycle-aware labels (e.g. preview browser “Clicking” / “Clicked” / failure states), **browser vs T3** icons (globe / wordmark), **browser-specific group counts** in summaries, and command summaries that show the real program after unwrapping bash/zsh wrappers while expanded rows still show the full command.
> 
> **Feed modeling** changes expanded groups from one timeline/feed row per call to a single `work-details` group row; live **shimmer** tracks actively running calls (including task progress) rather than any “live” summary. Mobile adds **Dynamic Type–aware** row sizing that drops fixed LegendList heights when accessibility text would overflow, and clears size caches on typography changes.
> 
> **Web timeline follow** no longer treats upward wheel/keyboard scroll inside a scrolled tool group as “leave the live edge,” via `toolGroupConsumesUpwardNavigation`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 53fa54843904d23dfe1ddb3aa7dc30ccd5e31590. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Improve chat tool group summaries, scroll restoration, and navigation
> - Expanded tool groups now render as one virtualized details list instead of separate per-entry rows, on both web ([MessagesTimeline.tsx](https://github.com/pingdotgg/t3code/pull/9106/files#diff-f2a34c4ad8d2b68c45657dbdcbf14afac4ea93e1fbd37007aaa2ccb8b41d2588)) and mobile ([thread-work-log.tsx](https://github.com/pingdotgg/t3code/pull/9106/files#diff-06f278f3ca8d6524ea378cdd0d5fea4e0a691986d150c0b93b43724ed8d40909)).
> - Added scroll-anchor save/restore and append-follow gating: expanded groups follow newly appended entries only when prior entries are unchanged and the reader is at the content edge; saved row and within-row offset are restored on re-render.
> - Recognized T3 MCP and browser tools now resolve to friendly lifecycle-aware labels and icons via shared resolvers in [presentation.ts](https://github.com/pingdotgg/t3code/pull/9106/files#diff-b6852793cd3916fd7bb7704e9bfe1391f7ca44403063a0502eb92d59f5060972); command summaries show program names with verbs like running, ran, failed, declined, and stopped.
> - Added `toolGroupConsumesUpwardNavigation` in [ChatView.logic.ts](https://github.com/pingdotgg/t3code/pull/9106/files#diff-71aecfcacbebc9adf2808573cb4e4ac389b319a6ce3fa79dccb8b87c6c26f59d) so wheel and keyboard navigation inside scrolled scrollable tool groups does not trigger timeline manual navigation.
> - Shell-wrapped commands in `commandProgramName` now resolve to the underlying program rather than the shell name ([commandLabel.ts](https://github.com/pingdotgg/t3code/pull/9106/files#diff-65081d51890ab0d6946a80f5a393cb7285813f75240ce83508d6ac9119f22c45)).
> - Risk: `deriveMessagesTimelineRows` and `appendToolGroupRows` now emit one details row per expanded group instead of one row per entry; in-tree readers and tests are updated, but any out-of-tree consumers expecting per-entry expanded rows must adopt the new details-group shape. `buildThreadFeed` copy text now uses normalized authored labels instead of friendly display labels.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 53fa548.</sup>
> 
>

### CODE REVIEW

**Comment by juliusmarminge** [comment]
Scroll-hardening follow-up in cc91b4d370d:

- Preserve group scroll position and expanded details across reopening and virtualization.
- Only follow actual appends at the hard end; preserve history-reading position and interrupted scrolling.
- Remove fixed-height assumptions for accessibility-scaled mobile rows and wait for short-group viewport layout before following an append.
- Reconcile a reproduced web bootstrap mismatch once at load, including a fault-injection check of the actual DOM offset.

[Full scroll verification recording](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/7b225fbf86d8760e/tool-group-scroll-verification.mp4). The PR body retains the original full walkthrough and individual before/after images.

363 focused tests, web/mobile typechecks, targeted lint, and diff checks passed. Web checks cover scroll position, follow behavior, and bounded rendering, not frame-time performance. Native gesture handoff, momentum, and large-font scrolling remain unverified; I recommend holding merge until that device pass.

**Comment by juliusmarminge** [comment]
Pushed `33638d0ef36` with fixes found during the native scroll pass:

- Shared offset-based anchors fix the five-row jump after a fast flick.
- Correct mobile rem-based heights and measured content sizes keep end following accurate.
- Font-only remeasurement fixes clipped native text without resetting the list during normal scrolling, streamed output, or disclosure changes.

371 focused tests passed, along with web/shared/mobile typechecks and targeted lint. React Doctor reported no errors and one existing chained-iteration warning outside the scroll path.

[Full native walkthrough, 107s](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/a9cc0cd3dca0c60f/native-scroll-verification.mp4) · [Final Dynamic Type round trip, 19s](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/c4a84bdc8933c459/native-dynamic-type-verification.mp4)

[Before/after: fast-flick restoration](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/97fe8c327ac1a6a7/native-restoration-comparison.png) · [Before/after: large text](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/4d00d3cb4c8cd344/native-large-text-comparison.png)

The PR description now includes these alongside the full web recordings and the original per-fix screenshots. iOS Simulator gestures passed; Android and physical-device frame timing are still unverified. The temporary mobile connection and services have been removed; the isolated web dev server remains running.

File: apps/web/src/components/ChatView.tsx [hunk 1]
```diff
@@ -4187,10 +4187,29 @@ function ChatViewContent(props: ChatViewProps) {
...
+            element && element !== group;
+            element = element.parentElement
+          ) {
+            if (element.scrollTop > 0) {
+              const overflowY = getComputedStyle(element).overflowY;
+              if (overflowY === "auto" || overflowY === "scroll") return true;
+            }
+          }
+          return false;
+        };
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/web/src/components/ChatView.tsx]: Fixed in 829ba9e0d1c.

The guard now checks actual `scrollTop > 0` on scrollable ancestors through the tool group, including nested output. The group no longer blocks native scroll chaining at its boundaries. The outer timeline keeps its containment.

This is web/desktop event handling. Mobile, list recycling, append following, and saved-anchor logic are unchanged.

Verification: 13 new boundary cases, 186 focused tests passing, web typecheck, targeted lint, and scoped React Doctor. In the browser, PageUp at the group's top moved the conversation from 181px to 0px and disabled follow. Three subsequent tool calls left it at 0px. ArrowUp inside nested output moved only the output and kept follow enabled.

| Before: upward input stays trapped | After: conversation scrolls upward |
| --- | --- |
| Before upward-scroll fix | After upward-scroll fix |

Full 7-second verification recording:

https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/9a5a0b7e72b5efc0/upward-scroll-escape.mp4

The browser pass uses native keyboard input and seeded in-memory UI data, not a provider run. The composer is hidden for privacy.

**Comment by juliusmarminge** [comment]
Pushed `3bc666665f6` with the four confirmed review fixes:

- Command summaries respect failed, stopped, completed, and declined states on web/desktop and mobile.
- Terminal `task.progress` updates no longer mark failed tasks as active work on either client.
- Web T3 tool icons use the shared, color-inheriting SVG wordmark.
- Failure marks precede the disclosure slot, keeping chevrons aligned with mobile.

214 focused tests, web and mobile typechecks, targeted lint, formatting, and diff checks passed. Scoped React Doctor found no errors; its two warnings concern the unchanged SVG path precision and existing row memoization.

| Fix | Before | After |
| --- | --- | --- |
| T3 icon tone and disclosure alignment | Before icon fixes | After icon fixes |
| Failed command summary | Failed command reported as Ran | Failed command reported as Failed |
| Failed task activity | Failed task shown as active | Failed task no longer active |

Full 24-second verification recording:

https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/f0eaa3de4f30a891/review-fixes.mp4

The new keyboard report did not reproduce with native input in the focused Electron/Chromium preview. PageUp and Home moved the conversation from 349px to 0px; ArrowUp moved it to 309px. All three disabled live-follow. Three subsequent tool calls preserved the history-reading position. The keyboard and scrolling code is unchanged in this commit.

https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/25ea65e33c8accbf/keyboard-boundary-verification.mp4

These web recordings use synthetic in-memory UI fixtures, not provider runs. The composer is hidden for privacy. Mobile lifecycle changes have unit and type coverage; this follow-up does not include another native-device pass. Earlier web and native recordings remain in the PR.

CI follow-up in `53fa5484390`: the existing completed-command component test still expected `Running pnpm`. It now expects `Ran pnpm` and rejects `Running pnpm`, matching the explicit completed lifecycle. The failure reproduced locally before this test-only correction. All 253 tests in the four focused timeline/lifecycle suites now pass, along with the web typecheck, targeted lint, formatting, and diff checks. No runtime or scrolling code changed.

File: apps/web/src/components/chat/MessagesTimeline.tsx [hunk 2]
```diff
@@ -2536,6 +2580,9 @@ const PlainWorkEntryRow = memo(function PlainWorkEntryRow(props: {
               )}
             />
           </span>
+          {showFailedIndicator && toolPresentation ? (
+            <XIcon aria-hidden className="size-3 shrink-0 text-icon-muted" />
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/web/src/components/chat/MessagesTimeline.tsx]: Fixed in `3bc666665f6`. The failure mark now comes before the fixed disclosure slot, so failed and successful rows keep the chevron in the same rightmost position. Mobile already used this order.

[Before](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/cf56a396b7d0a735/icons-before.png) / [after](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/280b6b3d39126dca/icons-after.png). Verified in the web preview, with targeted lint and typechecks passing.

File: apps/web/src/components/chat/MessagesTimeline.tsx [hunk 3]
```diff
@@ -2132,14 +2204,26 @@ type WorkEntryIconName =
...
+function WorkEntryIcon({ name, className }: { name: WorkEntryIconName; className: string }) {
   switch (name) {
     case "bot":
       return <BotIcon className={className} aria-hidden />;
+    case "browser":
+      return <GlobeIcon className={className} aria-hidden />;
+    case "t3-code":
+      return (
+        <img
+          src={`${import.meta.env.BASE_URL}apple-touch-icon.png`}
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/web/src/components/chat/MessagesTimeline.tsx]: Fixed in `3bc666665f6`. The sidebar's SVG wordmark is now shared with `WorkEntryIcon`, preserving `currentColor`, opacity, and the caller's tone classes. Browser tools still use Globe. Mobile already has a tint-aware wordmark.

[Before](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/cf56a396b7d0a735/icons-before.png) / [after](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/280b6b3d39126dca/icons-after.png). Verified in the web preview, with targeted lint and typechecks passing.

File: apps/mobile/src/lib/threadActivity.ts [hunk 4]
```diff
@@ -1512,40 +1524,42 @@ function appendToolGroupRows(
...
+  const presentation = resolveWorkEntryToolPresentation(
+    activity.workEntry,
+    active ? "inProgress" : "completed",
+  );
+  if (presentation) return presentation.displayName;
   const command = activity.workEntry.command?.trim();
   if (command) {
     const program = commandProgramName(command);
-    return program ? `Running ${program}` : "Running command";
+    const verb = active ? "Running" : activity.lifecycleStatus === "declined" ? "Declined" : "Ran";
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/mobile/src/lib/threadActivity.ts]: Fixed in `3bc666665f6` on mobile and web/desktop. Explicit command lifecycle status now determines the verb: Running, Ran, Failed, Declined, or Stopped. A still-running turn cannot override a terminal command's state.

Existing client test suites cover the outcomes, including shell-wrapped commands. [Before](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/20ffb86214e64dda/failed-before.png) / [after](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/c277e728f976f71d/failed-after.png) from the web verification.

File: apps/mobile/src/lib/threadActivity.ts [hunk 5]
```diff
@@ -1487,19 +1492,26 @@ function appendToolGroupRows(
...
   const expanded = expandedWorkGroupIds.has(groupId);
-  const latestInProgressActivity = activities.findLast(
+  const latestActiveActivity = activities.findLast(
     (activity) =>
-      isWorking && activity.lifecycleStatus === "inProgress" && activity.turnId === unsettledTurnId,
+      isWorking &&
+      activity.turnId === unsettledTurnId &&
+      (activity.lifecycleStatus === "inProgress" ||
+        (activeTail &&
+          (activity.workEntry.sourceActivityKind === "task.progress" ||
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/mobile/src/lib/threadActivity.ts]: Fixed in `3bc666665f6` on mobile and web/desktop. `task.progress` only falls back to active when lifecycle status is absent; explicit in-progress status remains active and terminal states do not.

Both client test suites cover the terminal cases. The web pass confirms the failed task no longer carries the active shimmer; ongoing turn work has its separate Thinking row. [Before](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/9a93efc34bfcf05c/task-before.png) / [after](https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/0d5bc945888e685c/task-after.png).

File: apps/web/src/components/ChatView.logic.ts [hunk 6]
```diff
@@ -99,6 +99,22 @@ export function shouldReleaseTimelineAnchorForToolActivity(input: {
...
+  // A nested result or the group itself can consume an upward scroll.
+  for (let element = elementTarget; element; element = element.parentElement) {
+    if (element.scrollTop > 0) {
+      const overflowY = getComputedStyle(element).overflowY;
+      if (overflowY === "auto" || overflowY === "scroll") return true;
+    }
+    if (element === group) break;
+  }
+  return false;
+}
```
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply apps/web/src/components/ChatView.logic.ts:116]: Not reproduced with native keyboard input in the focused Electron/Chromium preview, using the existing scroll code from `829ba9e`. No keyboard workaround is included in `3bc666665f6`.

With the tool group focused at `scrollTop = 0` and live-follow enabled:

| Key | Conversation offset | Live-follow afterward |
| --- | --- | --- |
| PageUp | 349px to 0px | disabled |
| Home | 349px to 0px | disabled |
| ArrowUp | 349px to 309px | disabled |

PageUp moved the group's rendered position from y=288 to y=637, confirming visible conversation movement as well as the DOM scroll offset. Three appended tool calls then left the conversation at 0px with follow disabled. Group focus was retained. The test used native input, not a dispatched KeyboardEvent.

https://gh-file-drop-api-prod-mi5fy3sowv63ufte.pinglabs.workers.dev/f/25ea65e33c8accbf/keyboard-boundary-verification.mp4

The fixture changes only in-memory conversation data; it does not patch keyboard or scroll handlers. Resolving this report based on the observed behavior. A reproduction in another browser engine would need its browser/version and exact focus steps.
